// @shopify/shopify-api/dist/cjs/rest/admin/2023-10/application_charge.js
'use strict';

var base = require('../../base.js');
var types = require('../../../lib/types.js');
var currency = require('./currency.js');

/***********************************************************************************************************************
* This file is auto-generated. If you have an issue, please create a GitHub issue.                                     *
***********************************************************************************************************************/
class ApplicationCharge extends base.Base {
    static apiVersion = types.ApiVersion.October23;
    static hasOne = {
        "currency": currency.Currency
    };
    static hasMany = {};
    static paths = [
        { "http_method": "get", "operation": "get", "ids": [], "path": "application_charges.json" },
        { "http_method": "get", "operation": "get", "ids": ["id"], "path": "application_charges/<id>.json" },
        { "http_method": "post", "operation": "post", "ids": [], "path": "application_charges.json" }
    ];
    static resourceNames = [
        {
            "singular": "application_charge",
            "plural": "application_charges"
        }
    ];
    static async find({ session, id, fields = null }) {
        const result = await this.baseFind({
            session: session,
            requireIds: true,
            urlIds: { "id": id },
            params: { "fields": fields },
        });
        return result.data ? result.data[0] : null;
    }
    static async all({ session, since_id = null, fields = null, ...otherArgs }) {
        const response = await this.baseFind({
            session: session,
            urlIds: {},
            params: { "since_id": since_id, "fields": fields, ...otherArgs },
        });
        return response;
    }
    confirmation_url;
    created_at;
    currency;
    id;
    name;
    price;
    return_url;
    status;
    test;
    updated_at;
}

exports.ApplicationCharge = ApplicationCharge;
