// @shopify/shopify-api/dist/cjs/runtime/index.js
'use strict';

var index = require('./http/index.js');
var types = require('./crypto/types.js');
var crypto = require('./crypto/crypto.js');
var utils$1 = require('./crypto/utils.js');
var runtimeString = require('./platform/runtime-string.js');
var cookies = require('./http/cookies.js');
var headers = require('./http/headers.js');
var utils = require('./http/utils.js');



Object.defineProperty(exports, "abstractConvertHeaders", {
	enumerable: true,
	get: function () { return index.abstractConvertHeaders; }
});
Object.defineProperty(exports, "abstractConvertIncomingResponse", {
	enumerable: true,
	get: function () { return index.abstractConvertIncomingResponse; }
});
Object.defineProperty(exports, "abstractConvertRequest", {
	enumerable: true,
	get: function () { return index.abstractConvertRequest; }
});
Object.defineProperty(exports, "abstractConvertResponse", {
	enumerable: true,
	get: function () { return index.abstractConvertResponse; }
});
Object.defineProperty(exports, "abstractFetch", {
	enumerable: true,
	get: function () { return index.abstractFetch; }
});
exports.isOK = index.isOK;
exports.setAbstractConvertHeadersFunc = index.setAbstractConvertHeadersFunc;
exports.setAbstractConvertIncomingResponseFunc = index.setAbstractConvertIncomingResponseFunc;
exports.setAbstractConvertRequestFunc = index.setAbstractConvertRequestFunc;
exports.setAbstractConvertResponseFunc = index.setAbstractConvertResponseFunc;
exports.setAbstractFetchFunc = index.setAbstractFetchFunc;
Object.defineProperty(exports, "HashFormat", {
	enumerable: true,
	get: function () { return types.HashFormat; }
});
Object.defineProperty(exports, "crypto", {
	enumerable: true,
	get: function () { return crypto.crypto; }
});
exports.setCrypto = crypto.setCrypto;
exports.asBase64 = utils$1.asBase64;
exports.asHex = utils$1.asHex;
exports.createSHA256HMAC = utils$1.createSHA256HMAC;
exports.hashString = utils$1.hashString;
Object.defineProperty(exports, "abstractRuntimeString", {
	enumerable: true,
	get: function () { return runtimeString.abstractRuntimeString; }
});
exports.setAbstractRuntimeString = runtimeString.setAbstractRuntimeString;
exports.Cookies = cookies.Cookies;
exports.addHeader = headers.addHeader;
exports.canonicalizeHeaderName = headers.canonicalizeHeaderName;
exports.canonicalizeHeaders = headers.canonicalizeHeaders;
exports.flatHeaders = headers.flatHeaders;
exports.getHeader = headers.getHeader;
exports.getHeaders = headers.getHeaders;
exports.removeHeader = headers.removeHeader;
exports.setHeader = headers.setHeader;
exports.splitN = utils.splitN;
